### DISCUSSION TITLE
Terragrunt Log Formatting

### DISCUSSION THREAD

**Discussion by yhakbar** [gruntwork-io/terragrunt#3558] [Pre-release Feedback]
Release [v0.68.14-beta2024111502](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.68.14-beta2024111502) introduces new controls for log formatting.

Share your feedback!

**Comment by yhakbar** [comment]
I'd especially like to invite @grimm26 , @lorengordon , @wyardley and @tide-jamiegwatkin to share your feedback.

We want to make sure we're addressing any concerns that you had early on during our logging redesign, and are soliciting any feedback you may have to improve logging customization further.

**Reply by wyardley** [reply]
Thanks! I am not in a position to test myself in the next few weeks for various reasons, but really excited to hear about this, and also glad to see it get some testing ahead of time.

**Reply by lorengordon** [reply]
Is there an env option to set `--terragrunt-log-custom-format`? I'm not seeing one mentioned in the [linked doc](https://github.com/gruntwork-io/terragrunt/blob/feat/custom-log-format/docs/_docs/02_features/custom-log-format.md)...

**Reply by lev-ok** [reply]
@lorengordon, Sure, there is [`terragrunt-log-custom-format`](https://github.com/gruntwork-io/terragrunt/blob/feat/custom-log-format/docs/_docs/04_reference/cli-options.md#terragrunt-log-custom-format).

**Comment by grimm26** [comment]
@yhakbar how can I use this to stop terragrunt from showing all errors twice?

**Reply by yhakbar** [reply]
Hey @grimm26 ,

This update has more to do with addressing issues like this:
https://github.com/gruntwork-io/terragrunt/issues/3431

And to give you full control over the formatting of logs.

We will shortly add another flag named `--terragrunt-log-disable-error-summary` to prevent emitting the summary of encountered errors after a `run-all` once the updates referenced here are released.

**Reply by yhakbar** [reply]
It wasn't that quick, but we do have an in-draft PR that has the code to address this:
https://github.com/gruntwork-io/terragrunt/pull/3687

We're currently in the process of redesigning our CLI and have a company-wide break coming up, so we might not be able to get it released soon. Just wanted to give a status update and make sure your concerns weren't being ignored.

**Comment by atheiman** [comment]
I'd like to run something like `terragrunt run-all plan -out tfplan --terragrunt-stdout-file terragrunt.stdout.txt` and send stdout for each directory to `<directory>/terragrunt.stdout.txt` and hide from stdout.

I use `run-all` against dozens / hundreds of directories and currently I get >300,000 lines of log output in my ci/cd interface. These logs are primarily excessively noisy Terraform `refreshing state ...` lines, which there are several issues on Terraform for quieting. Mostly the recommendation to quiet Terraform output is to run `terraform plan -out=tfplan >/dev/null`, however, i dont have a way to run that with `terragrunt run-all ...` (unless Im missing something).

**Reply by yhakbar** [reply]
You have two options:
1. You can use [terragrunt-out-dir](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-out-dir) to create a secondary directory to store all the output files that mirrors the filesystem of your stack.
2. You can set `-out` to `../../tfplan` in order to have the plan file stored next to the `terragrunt.hcl` files. By default, they'll be stored in the `.terragrunt-cache` directory.

Most users avoid trying to store the actual stdout of a run, and instead store the plan binary, then reproduce stdout as needed.

**Comment by atheiman** [comment]
Separately from my other comment... is there a value I can specify for `--terragrunt-log-custom-format` to only show the terraform command from `%tf-command-args` rather than the command with all the args? For example, `%tf-command-args` is currently expanding to `init -reconfigure -backend=false` but I'd rather just see `init` to know what stage of my init/plan/apply lifecycle the command is executing (I dont need to see `-reconfigure -backend=false` in the log prefix). Maybe a new replacement named `%tf-command` would be easier to use for this?

**Reply by yhakbar** [reply]
This sounds like a feature request!

If you have any golang experience, I don't imagine this would be a difficult task to tackle in a pull request to deliver the feature yourself.

**Reply by atheiman** [reply]
ok great, I just wanted to confirm there wasnt already a built-in replacement for this I was missing.

**Reply by atheiman** [reply]
new issue: #3691

**Comment by atheiman** [comment]
Also Ive noticed that stdout from `before_hook`within a given terragrunt dir is delayed until all `run-all` terraform command stdout is complete for that terragrunt dir. I have a `before_hook` that runs `terraform show tfplan | tf-summarize` to print the plan summary before it is applied (otherwise terraform doesnt tell you much about what it is doing, all you see is `<resource.path>: Modifying ...`)

Im not seeing the stdout from the `before_hook` until all the stdout from the module's stdout has been printed. So my before_hook to show plan summary on stdout is printed to the console after the apply output is printed to the console. This is in a ci/cd interface, so the terminal type may be having an effect.

**Reply by lev-ok** [reply]
> Also Ive noticed that stdout from `before_hook`within a given terragrunt dir is delayed until all `run-all` terraform command stdout is complete for that terragrunt dir. I have a `before_hook` that runs `terraform show tfplan | tf-summarize` to print the plan summary before it is applied (otherwise terraform doesnt tell you much about what it is doing, all you see is `<resource.path>: Modifying ...`)
> 
> Im not seeing the stdout from the `before_hook` until all the stdout from the module's stdout has been printed. So my before_hook to show plan summary on stdout is printed to the console after the apply output is printed to the console. This is in a ci/cd interface, so the terminal type may be having an effect.

This is done to ensure that output from different modules does not overlap each other, for example, if it is json, then mixing output will cause their corruption.

**Reply by yhakbar** [reply]
Terragrunt does buffer terminal output during `run-all` to make it easier to avoid mangling output (e.g. if you have one run emitting to stdout, you'd probably prefer that another one not cut it off mid-log).

If you have a simple fixture that can reliably reproduce the undesirable behavior, it might be easier to take a look and see if it's feasible to address.

I'm not sure if we can dedicate resources to it, however, as I'm not sure how many users this is impacting (and how badly). Even if it's evidently undesirable and easy to reproduce, we might need community contribution to address it (including the testing required to avoid accidentally mangling streams between runs of different units).

**Reply by atheiman** [reply]
Thank you both for each of the replies

@levkohimins Im good with the buffering of a single module output, I like that behavior. The issue Im running into is `before_hook` stdout is delayed until after the terraform command has run. The `before_hook` stdout should show before the terraform command stdout.

A simple example of the issue can be seen with `terragrunt run-all plan` with a subdirectory `terragrunt.hcl` with a `before_hook`

```hcl
before_hook "echo_to_stdout" {
  commands = ["plan"]
  execute = ["bash", "-c", "echo 'this should show before terraform STDOUT, but it is delayed until after terraform STDOUT'"]
}
```

**Comment by captainjack64** [comment]
I would like to request two things:

1. Allow Terragrunt to output to `stdout`, so we can pipe through external tools. I've opened a bug ticket here https://github.com/gruntwork-io/terragrunt/issues/3938
2. Ability to trim the `%msg` output to a set number of characters. The `width` argument does a similar thing but it pads the output with spaces, which I don't want. For example, if my `%msg` is over 10 chars long, my width set to 9 and suffix set to `...` then it looks like this:
```
"A long mes..."
```
However, if the `%msg` is under the `width` limit, then my output looks like this:
```
"Foo      ..."
```
...which doesn't look great.

3. Ability to capture Terragrunt CLI output to a file (or stdout as in 1 so we can `tee` it)
4. Remove blank lines from output. For example, the below:
```
terragrunt run-all plan --non-interactive
11:43:42.054 ERROR  Error: Error in function call


11:43:42.054 ERROR    on ./terragrunt.hcl line 6, in locals:

11:43:42.054 ERROR     6:     acc_id = get_aws_account_id()

11:43:42.054 ERROR

11:43:42.054 ERROR  Call to function "get_aws_account_id" failed: Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): SSOProviderInvalidToken: the SSO session has expired or is invalid.


11:43:42.054 ERROR  Error processing module at './terragrunt.hcl'. How this module was found: Terragrunt config file found in a subdirectory of ../playground. Underlying error: ./terragrunt.hcl:6,14-33: Error in function call; Call to function "get_aws_account_id" failed: Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): SSOProviderInvalidToken: the SSO session has expired or is invalid.
11:43:42.359 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
11:43:42.360 ERROR  Suggested fixes:
Missing AWS credentials. Provide credentials to proceed.
```
would look much better like this:
```
terragrunt run-all plan --non-interactive --detailed-exit-code
11:43:42.054 ERROR  Error: Error in function call
11:43:42.054 ERROR    on ./terragrunt.hcl line 6, in locals:
11:43:42.054 ERROR     6:     acc_id = get_aws_account_id()
11:43:42.054 ERROR
11:43:42.054 ERROR  Call to function "get_aws_account_id" failed: Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): SSOProviderInvalidToken: the SSO session has expired or is invalid.
11:43:42.054 ERROR  Error processing module at './terragrunt.hcl'. How this module was found: Terragrunt config file found in a subdirectory of ../playground. Underlying error: ./terragrunt.hcl:6,14-33: Error in function call; Call to function "get_aws_account_id" failed: Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): SSOProviderInvalidToken: the SSO session has expired or is invalid.
11:43:42.359 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
11:43:42.360 ERROR  Suggested fixes:
Missing AWS credentials. Provide credentials to proceed.
```
